pkg/fanal/licensing/classification/classifier.go

```diff
@@ -0,0 +1,127 @@
+package classification
+
+import (
+	"fmt"
+
+	"github.com/aquasecurity/trivy/pkg/fanal/types"
+
+	"github.com/go-enry/go-license-detector/v4/licensedb"
+	"github.com/google/licenseclassifier"
+	classifier "github.com/google/licenseclassifier/v2"
+	"github.com/google/licenseclassifier/v2/assets"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	IgnoredLicenses []string
+)
+
+type Classifier struct {
+	classifier *classifier.Classifier
+}
+
+func NewClassifier(ignoredLicenses []string) (*Classifier, error) {
+	var c *classifier.Classifier
+	IgnoredLicenses = ignoredLicenses
+
+	licensedb.Preload()
+
+	_, err := assets.ReadLicenseDir()
+	if err != nil {
+		return nil, err
+	}
+	c, err = assets.DefaultClassifier()
+	if err != nil {
+		return nil, err
+	}
+	return &Classifier{
+		classifier: c,
+	}, nil
+}
+
+// Classify detects and classifies the licensedFile found in a file
+func (c *Classifier) Classify(filePath string, contents []byte) (types.LicenseFile, error) {
+	licFile, err := c.googleClassifierLicense(filePath, contents)
+	if err != nil {
+		return licFile, err
+	}
+
+	if len(licFile.Findings) == 0 {
+		return c.fallbackClassifyLicense(filePath, contents)
+	}
+
+	return licFile, nil
+}
+
+func (c *Classifier) googleClassifierLicense(filePath string, contents []byte) (types.LicenseFile, error) {
+
+	license := types.LicenseFile{FilePath: filePath}
+	matcher := c.classifier.Match(c.classifier.Normalize(contents))
+
+	for _, m := range matcher.Matches {
+		riskLevel, classification := GoogleClassification(m.Name)
+		licenseLink := fmt.Sprintf("https://spdx.org/licenses/%s.html", m.Name)
+
+		license.Findings = append(license.Findings, types.LicenseFinding{
+			License:                          m.Name,
+			Confidence:                       m.Confidence,
+			GoogleLicenseClassificationIndex: riskLevel,
+			GoogleLicenseClassification:      classification,
+			LicenseLink:                      licenseLink,
+		})
+	}
+
+	return license, nil
+}
+
+func (c *Classifier) fallbackClassifyLicense(filePath string, contents []byte) (types.LicenseFile, error) {
+	license := types.LicenseFile{FilePath: filePath}
+
+	matcher := licensedb.InvestigateLicenseText(contents)
+	for l, confidence := range matcher {
+		riskLevel, classification := GoogleClassification(l)
+		licenseLink := fmt.Sprintf("https://spdx.org/licenses/%s.html", l)
+
+		license.Findings = append(license.Findings, types.LicenseFinding{
+			License:                          l,
+			Confidence:                       float64(confidence),
+			GoogleLicenseClassificationIndex: riskLevel,
+			GoogleLicenseClassification:      classification,
+			LicenseLink:                      licenseLink,
+		})
+	}
+
+	return license, nil
+}
+
+func GoogleClassification(licenseName string) (int, string) {
```

Comment:
@knqyf263 for my own understanding, can you clarify for me the requirement for identification and classification being separated?

Reply:
Sure. In image scanning, we basically want to leverage cache so that we can skip pulling the same layer next time. But license type classification such as "forbidden" and "restricted" can be modified by user (I'm implementing it now). In that case, we can't use the license type in cache. We need to re-classify the type according to the user's custom types. In other words, we don't have to store the license type in cache as we can't use it anyway.

1. Detect the license name such as MIT
  - It will be stored in cache
  - This information cannot be modified by user
2. Classify the license type such as "forbidden" according to the license name
  - The license name is pulled from cache
  - This type classification can be modified by user

That is why we have "analyzer" and "scanner". The analyzer stores static information in cache and a scanner performs dynamic detection based on that information in cache.

![image](https://user-images.githubusercontent.com/2253692/178447493-1b0d59f9-6093-4dea-9d9d-3023db93062d.png)

rather than

![image](https://user-images.githubusercontent.com/2253692/178447973-9a03cdaa-a511-4609-bebd-9479769e6dbe.png)

But misconfiguration and secret are a bit special because we didn't want to store file content in cache as it is large and may contain sensitive information. We also perform scanning in the analyzer.

In summary,
![image](https://user-images.githubusercontent.com/2253692/178451759-8d3cc17c-32d0-4cd0-9292-5b7d56474da7.png)

Reply:
Thanks, that makes a lot of sense. It helps solidify my understanding about the way the cache works - I had originally been passing the threshold value to the analyzer which meant once it was in the cache, threshold changes were ignored so I had to move the filtering to the report stage.

This is very helpful - thanks